include/jemalloc/internal/tcache_inlines.h

```diff
@@ -32,13 +32,65 @@ tcache_enabled_set(tsd_t *tsd, bool enabled) {
 	tsd_slow_update(tsd);
 }
 
+static inline unsigned
+thread_nhbins_get(tsd_t *tsd) {
+	return tsd_thread_nhbins_get(tsd);
+}
+
+static inline void
+thread_nhbins_set(tsd_t * tsd, unsigned thread_nhbins) {
+	if (thread_nhbins > TCACHE_NBINS_MAX) {
+		thread_nhbins = TCACHE_NBINS_MAX;
+	}
+	tsd_thread_nhbins_set(tsd, thread_nhbins);
+	size_t tcache_max = tsd_thread_tcache_max_get(tsd);
+	assert(thread_nhbins == sz_size2index(tcache_max) + 1);
+}
+
+static inline size_t
+thread_tcache_max_get(tsd_t *tsd) {
+	return tsd_thread_tcache_max_get(tsd);
+}
+
+static inline void
+thread_tcache_max_set(tsd_t *tsd, size_t tcache_max) {
+	/* tcache_max should not be larger than TCACHE_MAXCLASS_LIMIT. */
+	if (tcache_max > TCACHE_MAXCLASS_LIMIT) {
+		tcache_max = (size_t)TCACHE_MAXCLASS_LIMIT;
+	}
+	/* Ceil tcache_max to a valid size class. */
+	tcache_max = sz_s2u(tcache_max);
+
+	size_t old_tcache_max = tsd_thread_tcache_max_get(tsd);
+	if (old_tcache_max != tcache_max && tcache_available(tsd)) {
```

Comment:
Probably refactor this way (and include the tcache->arena setup), so that the core logic is not duplicated:
```
bool enabled = tcache_available(tsd);
arena_t *assigned_arena;

if (enabled) {
  assigned_arena = tcache_slow->arena;
  /* Shutdown and reboot the tcache for a clean slate. */
  tcache_cleanup(tsd);
}
tsd_thread_tcache_max_set(tsd, tcache_max);
thread_nhbins_set(tsd, sz_size2index(tcache_max) + 1);
tsd_slow_update(tsd);
if (enabled) {
  tsd_tcache_data_init(tsd);
  tcache_arena_associate(... assigned_arena);
}
```